### PR TITLE
Stacey dev

### PR DESCRIPTION
**Pull request by staceywhitmore-inl**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
#2173

##### What are the significant changes in functionality due to this change request?
Run now completes successfully with samples.csv output. Updated finalizeCodeOutput() method to return SQL query info (in Pandas DataFrame) as dictionary (where each key is the column name queried from DataBase). Also, Moved abce_Custom_Sampler.xml to /tests/framework/CodeInterfaceTests. /abcesetting and samples.csv output were moved there as well.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by staceywhitmore-inl** [comment]
Run now completes successfully with samples.csv output. Updated finalizeCodeOutput() method to return SQL query info (in Pandas DataFrame) as dictionary (where each key is the column name queried from DataBase). Also, Moved abce_Custom_Sampler.xml to /tests/framework/CodeInterfaceTests. /abcesetting and samples.csv output were moved there as well.

File: ABCE_XML_BACKUPS/abce_Custom_Sampler.xml [hunk 1]
```diff
@@ -0,0 +1,115 @@
+<?xml version="1.0" ?>
```
Comment (joshua-cogliati-inl) [inline ABCE_XML_BACKUPS/abce_Custom_Sampler.xml:1]: The files in ABCE_XML_BACKUPS and OUT should be removed.

File: .gitignore [hunk 2]
```diff
@@ -71,3 +71,22 @@ plugins/plugin_directory.xml
 # vscode
 .vscode
 *.code-workspace
+
+# Custom
+STACEY_NOTES.md
+./OUT/*
+OUT
```
Comment (joshua-cogliati-inl) [inline .gitignore:78]: Adding OUT is probably okay, but the rest of the additions in this file are kind of specific and probably shouldn't be in the pull request.

File: ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+import os
+import re
+import warnings
+import pandas as pd
+import sqlite3
+
+from ravenframework.CodeInterfaceBaseClass import CodeInterfaceBase
+from ..Generic.GenericCodeInterface import GenericParser
+
+# For Memoization testing (only benefitial if same args are sometimes passed to finalizeCodeOutput)
```
Comment (joshua-cogliati-inl) [inline ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py:28]: Hm, if we want to do Memoization, it probably needs to be under user control, and should probably be done at a higher level than in an individual CodeInterface.

File: ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+      @ In, fargs, dict, optional, a dictionary containing the auxiliary input file variables the user can specify in the input (e.g. under the node < Code >< fileargstype =0 input0arg =0 aux0extension =0 .aux0/ >< /Code >)
+      @ In, preExec, string, optional, a string the command that needs to be pre-executed before the actual command here defined
+      @ Out, returnCommand, tuple, tuple containing the generated command. returnCommand[0] is the command to run the code (string), returnCommand[1] is the name of the output root
+    """
+    if clargs==None:
+      raise IOError('No input file was specified in clargs!')
+    #check for duplicate extension use
+    usedExts = list(ext[0][0] for ext in clargs['input'].values() if len(ext) != 0)
+    if len(usedExts) != len(set(usedExts)):
+      raise IOError('GenericCodeInterface cannot handle multiple input files with the same extension.  You may need to write your own interface.')
```
Comment (joshua-cogliati-inl) [inline ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py:54]: GenericCodeInterface should be changed to be specific to Abce

File: ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+          if 'scenario_name' in line:
+            scenarioName = line.split(':')[1].strip()
+            # remove the double quotes
+            scenarioName = scenarioName.replace('"','')
+            break
+      self._outputDirectory = os.path.join(os.path.dirname(settingsFile.getAbsFile()),'outputs',scenarioName)
+      return None
+
+    #prepend
+    todo = ''
```
Comment (joshua-cogliati-inl) [inline ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py:96]: There are some improvements that have been added to https://github.com/idaholab/raven/pull/2151 that need to be merged into this file.

File: ravenframework/Models/Code.py [hunk 6]
```diff
@@ -772,13 +772,13 @@ def collectOutput(self,finishedJob,output,options=None):
 
     self._replaceVariablesNamesWithAliasSystem(evaluation, 'input',True)
     # in the event a batch is run, the evaluations will be a dict as {'RAVEN_isBatch':True, 'realizations': [...]}
+    # breakpoint() # check value of evaluation variable
```
Comment (joshua-cogliati-inl) [inline ravenframework/Models/Code.py:775]: Both changes in this file can probably be reverted.

File: tests/framework/CodeInterfaceTests/abce_Custom_Sampler.xml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  <RunInfo>
+    <JobName>abceGCI</JobName>
+    <Sequence>sweep</Sequence>
+    <WorkingDir>abcesetting</WorkingDir>
+    <batchSize>1</batchSize>
+  </RunInfo>
+
+  <Steps>
+    <MultiRun name="sweep">
+      <!-- <Input class="Files" type="">abce_runner.sh</Input> -->
```
Comment (joshua-cogliati-inl) [inline tests/framework/CodeInterfaceTests/abce_Custom_Sampler.xml:12]: Old commented out lines should usually be removed unless there is a reason to keep them.

File: tests/framework/CodeInterfaceTests/tests [hunk 8]
```diff
@@ -139,6 +139,13 @@
    csv = 'PrescientInterfaceTest/out_prescient_fake_0.csv PrescientInterfaceTest/out_prescient_fake_1.csv'
    text = 'PrescientInterfaceTest/runPrescientFake/1/deterministic_with_network_scenarios_day/pyspdir_twostage/2020-07-10/Scenario_actuals.dat PrescientInterfaceTest/runPrescientFake/2/deterministic_with_network_scenarios_day/pyspdir_twostage/2020-07-10/Scenario_actuals.dat '
    test_interface_only = True
-   
  [../]
+
+ [./s_whitmoreTest]
```
Comment (joshua-cogliati-inl) [inline tests/framework/CodeInterfaceTests/tests:144]: This could use a name that reflects that this tests abce.

**Review by joshua-cogliati-inl — CHANGES_REQUESTED** [review]
This is making progress, but here are various things to consider fixing.

File: tests/framework/CodeInterfaceTests/tests (hunk 8 above)
Comment (staceywhitmore-inl) [reply tests/framework/CodeInterfaceTests/tests:144]: Good point. Let me know if you have any suggestions or ideas.

File: .gitignore (hunk 2 above)
Comment (staceywhitmore-inl) [reply .gitignore:78]: I had added OUT to the .gitignore but it looks like it was still added to the commit. I will take another look.

File: ABCE_XML_BACKUPS/abce_Custom_Sampler.xml (hunk 1 above)
Comment (staceywhitmore-inl) [reply ABCE_XML_BACKUPS/abce_Custom_Sampler.xml:1]: Agreed. I had added them to the .gitignore but they still got added to the commit.

File: .gitignore (hunk 2 above)
Comment (staceywhitmore-inl) [reply .gitignore:78]: STACEY_NOTES.md should also have been in the .gitignore. Apologies for not catching this before pushing.

File: ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py (hunk 4 above)
Comment (staceywhitmore-inl) [reply ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py:54]: OK. How do you mean? Are you just referring to the error message? (I haven't looked at some of these other methods stubs as much.)

File: ravenframework/CodeInterfaceClasses/Factory.py [hunk 9]
```diff
@@ -19,6 +19,7 @@
 
 from ..CodeInterfaceBaseClass import CodeInterfaceBase
 from .Generic.GenericCodeInterface import GenericCode
+from .Abce.AbceCodeInterface import Abce # NEW
```
[the remark this answers is not in the document]
Comment (staceywhitmore-inl) [reply ravenframework/CodeInterfaceClasses/Factory.py:22]: Forgot I had added that. Comment Removed.

File: ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    """
+      Convert csv (ADDENDUM: NOT csv But should be "convert SQlite...") information to RAVEN's prefered formats [Pandas DataFrame]
+      Joins together two different csv [again, NOT csv But SQLite]files and also reorders it a bit.
+      @ In, command, ignored  [Ignore for now]
+      @ In, codeLogFile, ignored [Ignore for now]
+      @ In, subDirectory, string, the subdirectory where the information is. [Use full path]
+      @ Out, directory, string, the assets results
+    """
+    # Cannot test if caching/memoization is needed until f(n) parameters are decided on.
+    # TODO: [First] Find Out: if we inted to use the Command, codeLogFile, and subDirectory parameters
```
[the remark this answers is not in the document]
Comment (staceywhitmore-inl) [reply ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py:162]: Removed most comments.

File: ravenframework/Models/Code.py (hunk 6 above)
Comment (joshua-cogliati-inl) [reply ravenframework/Models/Code.py:775]: Removing of a newline.

File: ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py (hunk 4 above)
Comment (joshua-cogliati-inl) [reply ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py:54]: Yes, mostly just the error message.

File: tests/framework/CodeInterfaceTests/tests (hunk 8 above)
Comment (staceywhitmore-inl) [reply tests/framework/CodeInterfaceTests/tests:144]: I've named it abceTest for now.

File: ravenframework/Models/Code.py [hunk 11]
```diff
@@ -772,13 +772,13 @@ def collectOutput(self,finishedJob,output,options=None):
...
     if isinstance(evaluation,dict) and evaluation.get('RAVEN_isBatch',False):
       for rlz in evaluation['realizations']:
         output.addRealization(rlz)
     # otherwise, we received a single realization
     else:
       output.addRealization(evaluation)
-
```
Comment (joshua-cogliati-inl) [inline ravenframework/Models/Code.py:781]: This line probably does not need to be removed.

**Review by staceywhitmore-inl — COMMENTED** [review]
Samples csv and xml output to added /gold. Cleaned out comments and removed prior, unneeded output files (pushed before updating .gitignore exclusions). Runs complete successfully.

**Comment by joshua-cogliati-inl** [comment]
This was merged into #2184  and so can be closed.

**Pull request closed without merging**